### PR TITLE
feat: Prune old resources

### LINKED ISSUE #88: Pruning resources
At present, Tanka will not remove Kubernetes resources that have been removed from the respective Jsonnet configuration, and such removals need to be done manually.

Terraform (for example) handles this by maintaining a record of 'intended state'. If an object is present in the state file, but not in the Jsonnet configuration, then we can deduce that it needs to be removed.

Given that Kubernetes already maintains its own state, and that that is consumed for `ks diff` etc, we could likely achieve deletions if we simply stored a list of resources created by this Tanka environment, within for example, a single CRD. This CRD could be named within the `spec.json` file.

Thus, when either `tk diff` or `tk apply` are executed, Tanka can look for any resources that are present in this CRD but are not present in the Jsonnet output. Such resources are candidates for deletion. `tk diff` could include them in its output, and `tk apply` could use `kubectl delete` to remove them.

We should consider whether deletions should be a standard feature of Tanka, or whether they would need enabling via a command line switch.

### PR DESCRIPTION
**Pull request by sh0rez**
Implements a mechanism that allows to remove previously created resources that are no longer present in the Jsonnet code, based on internal discussion and tips from @anguslees (thanks!)

Process:
1. Automatically labels all resources with `tanka.dev/environment: <env-name>`
2. Use `tk prune` to remove old ones
3. Fetches all `metadata.uid` of currently present objects
4. Fetches all objects that hold our label
5. Shows diff and prompts for confirm
6. Removes all objects having an unknown UID

Design doc: https://docs.google.com/document/d/1R3_dvXRb2d46GJTb3PQT3hfx2q5sy-ixifSGaWwDM8M

Fixes #88 

/cc @malcolmholmes @captncraig @mplzik

### CODE REVIEW

File: pkg/kubernetes/apply.go [hunk 1]
```diff
@@ -2,35 +2,98 @@ package kubernetes
...
+			continue
+		}
+
+		// ComponentStatus resource is broken in Kubernetes versions
+		// below 1.17, it will be returned even if the label does not
+		// match. Ignoring it here is fine, as it is an internal object
+		// type.
+		if m.APIVersion() == "v1" && m.Kind() == "ComponentStatus" {
+			continue
+		} else if m.APIVersion() == "v1" && m.Kind() == "Endpoints" {
```
Comment (woodsaj) [inline pkg/kubernetes/apply.go:74]: There are use-cases where you might need to create endpoints directly. 
https://kubernetes.io/docs/concepts/services-networking/service/#without-selectors

Though as endpoints dont include any metadata to indicate they are managed by a controller, i am not sure how we would ever know that was the case.

Comment (sh0rez) [reply pkg/kubernetes/apply.go:74]: True, but `Endpoint` has `ObjectMeta` and thus `metadata.labels` – which we can identify those we did create ourselves – right?

Comment (anguslees) [reply pkg/kubernetes/apply.go:74]: Unfortunately, automatically-created Endpoints get all the labels from the original Service.  They are the only core object where this happens (it's probably a bug).

Comment (anguslees) [reply pkg/kubernetes/apply.go:74]: Possible approaches:
- just always delete all Endpoints ([qbec does this](https://github.com/splunk/qbec/blob/65ec57bed31ad5bf3fde654695b6b05918b2dff6/internal/remote/client.go#L268))
- use annotations not labels to mark objects ([kubecfg does this](https://github.com/bitnami/kubecfg/blob/9aa6e1279ba939f98b4430b3af099f86ab6011b5/pkg/kubecfg/update.go#L483) although is migrating to labels)
- use both annotation and label (kubecfg will probably do this in future)
- use some other info to determine the original groupkind under which the object was created (kubecfg might also do this - using the serialised info in last-applied-configuration annotation).   This has some impact on migrated objects (eg extensions -> apps) too that should be thought through.
- add the original groupkind into the labels too - so we can ignore objects that have "changed kinds" in the server-side filter (I only just thought of this).  Same caveat as above re migrated apigroups.

File: pkg/kubernetes/apply.go [hunk 2]
```diff
@@ -2,35 +2,98 @@ package kubernetes
...
+		// ignore known ones
+		if uids[m.Metadata().UID()] {
+			continue
+		}
+
+		// ComponentStatus resource is broken in Kubernetes versions
+		// below 1.17, it will be returned even if the label does not
+		// match. Ignoring it here is fine, as it is an internal object
+		// type.
+		if m.APIVersion() == "v1" && m.Kind() == "ComponentStatus" {
```
Comment (anguslees) [inline pkg/kubernetes/apply.go:72]: Instead of hard-coding ComponentStatus here, maybe you could just repeat the label check (with a comment explaining why this was necessary for at least the ComponentStatus bug)

File: pkg/kubernetes/apply.go (hunk 1 above)
Comment (sh0rez) [reply pkg/kubernetes/apply.go:74]: Because we use `kubectl` for every `apply`, everything manually created should have the `kubectl.kubernetes.io/last-applied-configuration` annotation, right? So I guess it's sufficient to test for its existence?

Comment (sh0rez) [reply pkg/kubernetes/apply.go]: Implemented that, seems to work

File: pkg/kubernetes/reconcile.go [hunk 3]
```diff
@@ -52,21 +52,29 @@ var kindOrder = []string{
...
+func Reconcile(raw map[string]interface{}, cfg v1alpha1.Config, targets []*regexp.Regexp) (state manifest.List, err error) {
 	extracted, err := extract(raw)
 	if err != nil {
 		return nil, errors.Wrap(err, "flattening manifests")
 	}
 
 	out := make(manifest.List, 0, len(extracted))
 	for _, m := range extracted {
+		// inject tanka.dev/environment label
+		m.Metadata().Labels()[LabelEnvironment] = cfg.Metadata.NameLabel()
```
Comment (woodsaj) [inline pkg/kubernetes/reconcile.go:76]: This needs to be behind a feature flag or enabled through the tanka Config.
We dont want labels to suddenly start being added as soon as one person updates their tanka version.

Comment (sh0rez) [reply pkg/kubernetes/reconcile.go]: done

File: docs/docs/config.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+route: "/config"
+---
+
+# Configuration Reference
+
+Tanka's behavior can be customized per Environment using a file called `spec.json`
+
+## File format
+
+```json
```
Comment (tragiclifestories) [inline docs/docs/config.md:12]:
```suggestion
```js
```

Just a drive-by suggestion from a random person on the internet 😅 Because comments are not valid JSON, Github really doesn't seem to like rendering this block with a JSON syntax specified ...

Screenshot 2020-04-20 at 12 29 10

Comment (sh0rez) [reply docs/docs/config.md:12]: True, it looks ugly on GitHub, but renders better on our website when having it as json instead of js. The website is powered by the vscode highlight engine, which deals fine with this:

https://deploy-preview-251--tanka.netlify.app/config#file-format

Screen Shot 2020-04-20 at 14 50 21
